trick-ICG does not ignore forward declarations after Trick 13.5

trick-ICG creates io_src.cpp files that are not compilable when a forward declaration is not in the same file as file as the declaration as in the example below.

File 1: example_types.h
```c
typedef struct data_ data;

```

File 2: example.h
```c
#include "example_types.h"

struct data_ {
  double test;
};

```

When trick-ICG is run on example.h, the file io_example_types.cpp will not compile.  I found this when trying to created a trickified library, it gives the following.

```
at: build/S_library_swig: No such file or directory
Compiling  build/home/jmgentil/test/io_example_types.cpp
Compiling  build/home/jmgentil/test/io_example.cpp
build/home/jmgentil/test/io_example_types.cpp: In function ‘size_t io_src_sizeof_data()’:
build/home/jmgentil/test/io_example_types.cpp:55:23: error: invalid application of ‘sizeof’ to incomplete type ‘data’
     return sizeof(data) ;
                       ^
build/home/jmgentil/test/io_example_types.cpp: In function ‘void* io_src_allocate_data(int)’:
build/home/jmgentil/test/io_example_types.cpp:59:48: error: invalid application of ‘sizeof’ to incomplete type ‘data’
     data* temp = (data*)calloc(num, sizeof(data));
                                                ^
make[1]: *** [build/home/jmgentil/test/io_example_types.o] Error 1
make[1]: *** Waiting for unfinished jobs....
g++: warning: /home/jmgentil/test: linker input file unused because linking not done
make: *** [all] Error 2
```
I believe this is because it should ignore typedef struct data_ data which is declared in example_types.h.  This is what it did in Trick 13.5 and I did not see this issue.  I traced the change to `trick_source/codegen/Interface_Code_Gen/TranslationUnitVisitor.cpp`.  Changing the following section of code corrected the issue.

```cpp
// Had to add line below to match ICG from Trick 13.5 to get forward declarations to work properly
//            if ( rd != NULL ) {
            if ( rd != NULL and ! getFileName(ci , crd->getRBraceLoc(), hsd).empty() ) {
                std::string rd_file = getFileName(ci , rd->RBRACELOC(), hsd) ;
                std::string crd_file = getFileName(ci , crd->RBRACELOC(), hsd) ;
                if (!crd_file.empty() and !crd_file.compare(rd_file)) {
                    //crd->dump() ; std::cout << std::endl ;
                    if ( isInUserCode(ci , crd->RBRACELOC(), hsd) ) {
                        CXXRecordVisitor cvis(ci , cs, hsd , pa, true) ;

                        cvis.TraverseCXXRecordDecl(static_cast<clang::CXXRecordDecl *>(d)) ;
                        pa.printClass(cvis.get_class_data()) ;

                        /* Check to see if the struct/class is forward declared in the same file.
                           If it is, then remove the notation that it is forward declared.  This
                           is to allow C structs to be forward declared and typedeffed and io_src
                           code will be generated for both the original structure name and typedeffed
                           name.

                           struct Astruct ;
                           typedef struct Astruct {} Bstruct ;
                        */
                        std::set< std::string >::iterator it ;
                        std::string file_name = getFileName(ci , d->getLocEnd(), hsd) ;
                        std::string source_type = cvis.get_class_data()->getName() ;
                        it = fwd_declared_classes[file_name].find(source_type) ;
                        if ( it != fwd_declared_classes[file_name].end() ) {
                            fwd_declared_classes[file_name].erase(it) ;
                        }
                    }
                }
            } else {
                // These are forward declarations.  Insert this into the set of fwd declares keyed by the current file.
                if ( ! crd->getNameAsString().empty() ) {
                    fwd_declared_classes[getFileName(ci , d->getLocEnd(), hsd)].insert(crd->getNameAsString()) ;
                }
            }
```

Please advise if this is truly a bug or if I am doing something incorrectly.

--- a/trick_source/codegen/Interface_Code_Gen/TranslationUnitVisitor.cpp
+++ b/trick_source/codegen/Interface_Code_Gen/TranslationUnitVisitor.cpp
@@ -40,46 +40,56 @@ bool TranslationUnitVisitor::TraverseDecl(clang::Decl *d) {
         break ;
         case clang::Decl::CXXRecord : {
             clang::CXXRecordDecl * crd = static_cast<clang::CXXRecordDecl *>(d) ;
+            clang::RecordDecl * rd = crd->getDefinition() ;
+            bool is_forward_declaration = false;
+
             /* The definition of the record must exist before we can process it.  The definition is
                NULL when this is only a forward declaration of a class */
+            if ( rd == NULL ) {
+                is_forward_declaration = true;
+            }
             /* Sometimes rd is not NULL when we still only have a forward declaration because
                the file that defines the class is included before processing this fwd declare.
                In this case the CXXRecordDecl file name will not match the current file name, and is
                in fact empty */
-            clang::RecordDecl * rd = crd->getDefinition() ;
-            if ( rd != NULL ) {
+            else {
                 std::string rd_file = getFileName(ci , rd->RBRACELOC(), hsd) ;
                 std::string crd_file = getFileName(ci , crd->RBRACELOC(), hsd) ;
-                if (!crd_file.empty() and !crd_file.compare(rd_file)) {
-                    //crd->dump() ; std::cout << std::endl ;
-                    if ( isInUserCode(ci , crd->RBRACELOC(), hsd) ) {
-                        CXXRecordVisitor cvis(ci , cs, hsd , pa, true) ;
-
-                        cvis.TraverseCXXRecordDecl(static_cast<clang::CXXRecordDecl *>(d)) ;
-                        pa.printClass(cvis.get_class_data()) ;
-
-                        /* Check to see if the struct/class is forward declared in the same file.
-                           If it is, then remove the notation that it is forward declared.  This
-                           is to allow C structs to be forward declared and typedeffed and io_src
-                           code will be generated for both the original structure name and typedeffed
-                           name.
-
-                           struct Astruct ;
-                           typedef struct Astruct {} Bstruct ;
-                        */
-                        std::set< std::string >::iterator it ;
-                        std::string file_name = getFileName(ci , d->getLocEnd(), hsd) ;
-                        std::string source_type = cvis.get_class_data()->getName() ;
-                        it = fwd_declared_classes[file_name].find(source_type) ;
-                        if ( it != fwd_declared_classes[file_name].end() ) {
-                            fwd_declared_classes[file_name].erase(it) ;
-                        }
-                    }
+                if (crd_file.empty() || crd_file.compare(rd_file)) {
+                    is_forward_declaration = true;
                 }
-            } else {
+            }
+
+            if (is_forward_declaration) {
                 // These are forward declarations.  Insert this into the set of fwd declares keyed by the current file.
                 if ( ! crd->getNameAsString().empty() ) {
                     fwd_declared_classes[getFileName(ci , d->getLocEnd(), hsd)].insert(crd->getNameAsString()) ;
+                }
+            }
+            else {
+                //crd->dump() ; std::cout << std::endl ;
+                if ( isInUserCode(ci , crd->RBRACELOC(), hsd) ) {
+                    CXXRecordVisitor cvis(ci , cs, hsd , pa, true) ;
+
+                    cvis.TraverseCXXRecordDecl(static_cast<clang::CXXRecordDecl *>(d)) ;
+                    pa.printClass(cvis.get_class_data()) ;
+
+                    /* Check to see if the struct/class is forward declared in the same file.
+                       If it is, then remove the notation that it is forward declared.  This
+                       is to allow C structs to be forward declared and typedeffed and io_src
+                       code will be generated for both the original structure name and typedeffed
+                       name.
+
+                       struct Astruct ;
+                       typedef struct Astruct {} Bstruct ;
+                    */
+                    std::set< std::string >::iterator it ;
+                    std::string file_name = getFileName(ci , d->getLocEnd(), hsd) ;
+                    std::string source_type = cvis.get_class_data()->getName() ;
+                    it = fwd_declared_classes[file_name].find(source_type) ;
+                    if ( it != fwd_declared_classes[file_name].end() ) {
+                        fwd_declared_classes[file_name].erase(it) ;
+                    }
                 }
             }
         }

